src/session_state.rs

```diff
@@ -4,27 +4,33 @@
 use embedded_nal::IpAddr;
 use heapless::{String, Vec};
 
-pub struct SessionState {
+use embedded_time::{duration::Extensions, Clock, Instant};
+
+pub struct SessionState<C: Clock> {
     // Indicates that we are connected to a broker.
     pub connected: bool,
-    pub keep_alive_interval: u16,
+    pub keep_alive_interval: Option<u16>,
+    pub ping_timeout: Option<Instant<C>>,
     pub broker: IpAddr,
     pub maximum_packet_size: Option<u32>,
     pub client_id: String<64>,
+    last_transmission: Option<Instant<C>>,
     pub pending_subscriptions: Vec<u16, 32>,
     packet_id: u16,
     active: bool,
 }
 
-impl SessionState {
-    pub fn new<'a>(broker: IpAddr, id: String<64>) -> SessionState {
+impl<C: Clock> SessionState<C> {
+    pub fn new<'a>(broker: IpAddr, id: String<64>) -> SessionState<C> {
         SessionState {
             connected: false,
             active: false,
+            ping_timeout: None,
             broker,
             client_id: id,
             packet_id: 1,
-            keep_alive_interval: 0,
+            keep_alive_interval: Some(60),
```

Comment:
Any broker-provided configuration will always override any value here (or user specified). This is just some value to use to allow detection to occur. With the current value, we will ping the broker at 30s intervals if there's no out-bound traffic otherwise, which I think is a reasonable amount of time to recover a lost connection, but also not clog the network. Open to further opinions though